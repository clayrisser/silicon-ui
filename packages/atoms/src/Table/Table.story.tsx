import React from 'react';
import { withKnobs, text, select } from '@storybook/addon-knobs';
import Table from './Table';
import TableRow from '../TableRow';
import TableData from '../TableData';
import { StyleSheet, View } from 'react-native';

import { Rows } from 'react-native-table-component';

import Wrapper from '../../storybook/Wrapper';
import storiesOf from '../../storybook/storiesOf';
import themes from '../themes';
import withThemesProvider from '../../storybook/withThemesProvider';
// import docs from './Table.docs.mdx';

const tableHead = ['Head', 'Head2', 'Head3', 'Head4'];

// const tableData = [
//   { id: '256', name: 'kiran', test: 'test', etc: 'etc' },
//   { id: '256', name: 'kiran', test: 'test', etc: 'etc' },
//   { id: '256', name: 'kiran', test: 'test', etc: 'etc' }
// ];
const tableData = [
  ['1', '2', '3', '4'],
  ['a', 'b', 'c', 'd'],
  ['1', '2', '3', '456'],
  ['a', 'b', 'c', 'd']
];

storiesOf('Table', module)
  .addDecorator(withKnobs)
  .addDecorator(withThemesProvider(themes))
  .addParameters({
    // docs: { page: docs },
    // jest: ['Table.spec.tsx']
  })
  .add('with knobs', () => (
    <Wrapper>
      <Table
        backgroundColor={text('backgroundColor', 'secondary')}
        // borderRadius={number('borderRadius', 0)}
        // height={200}
        // width={200}
        autoContrast={select<'A' | 'AA' | 'AAA'>(
          'autoContrast',
          {
            false: '' as 'A',
            A: 'A',
            AA: 'AA',
            AAA: 'AAA'
          },
          'A'
        )}
      >
        <TableRow
          data={tableHead}
          thStyles={{ border: '1px solid #dddddd', padding: '8px' }}
        />
        <TableData
          data={tableData}
          tdStyles={{ border: '1px solid #dddddd', padding: '8px' }}
        />
      </Table>
    </Wrapper>
  ));

const styles = StyleSheet.create({});
