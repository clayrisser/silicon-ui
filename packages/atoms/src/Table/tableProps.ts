import { DetailedHTMLProps, TableHTMLAttributes, ReactNode } from 'react';
import { Table } from 'react-native-table-component';
// import { NativeBase } from 'native-base';
import {
  BackgroundProps,
  BorderProps,
  ColorProps,
  LayoutProps,
  PositionProps,
  ShadowProps,
  SpaceProps,
  TypographyProps
} from 'styled-system';
import { Theme } from '../themes';

export type DetailedHTMLTableProps = DetailedHTMLProps<
  TableHTMLAttributes<HTMLTableElement>,
  HTMLTableElement
>;

export interface StyledTableProps
  extends BackgroundProps,
    BorderProps,
    ColorProps,
    LayoutProps,
    PositionProps,
    ShadowProps,
    SpaceProps,
    TypographyProps {}

export interface NativeTableProps extends Table {}

export interface NativeItemProps extends Omit<Table, 'style'> {}

export interface CustomTableProps {
  autoContrast?: boolean | 'A' | 'AA' | 'AAA';
  children?: ReactNode;
  style?: any;
  theme?: Theme;
}

export interface TableProps
  extends CustomTableProps,
    NativeTableProps,
    NativeItemProps,
    StyledTableProps {}

export const customTablePropsKeys = new Set([
  'autoContrast',
  'children',
  'style',
  'theme'
]);

export const nativeTablePropsKeys = new Set<string>(['borderStyle', 'style']);

export const nativeItemPropKeys = new Set([
  'bordered',
  'disabled',
  'error',
  'underline'
]);

export interface SplitProps {
  customTableProps: CustomTableProps;
  nativeTableProps: NativeTableProps;
  nativeItemProps: NativeItemProps;
  styledTableProps: StyledTableProps;
}

export function splitProps(props: TableProps): SplitProps {
  const styledTableProps: { [key: string]: any } = {};
  const customTableProps: { [key: string]: any } = {};
  const nativeTableProps: { [key: string]: any } = {};
  const nativeItemProps: { [key: string]: any } = {};
  Object.entries({ ...props }).forEach(([key, prop]: [string, any]) => {
    if (customTablePropsKeys.has(key)) {
      customTableProps[key] = prop;
    } else if (nativeTablePropsKeys.has(key)) {
      nativeTableProps[key] = prop;
    } else if (nativeItemPropKeys.has(key)) {
      nativeItemProps[key] = prop;
    } else {
      styledTableProps[key] = prop;
    }
  });
  return {
    customTableProps,
    nativeTableProps,
    nativeItemProps,
    styledTableProps
  };
}

export const antiForwardTablePropsKeys = new Set<string>([]);
