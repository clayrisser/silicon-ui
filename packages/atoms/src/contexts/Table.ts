import { createContext, SetStateAction, Dispatch } from 'react';
import { Col } from '../types';

export type TableContextValue = [
  TableMeta | null,
  Dispatch<SetStateAction<TableMeta | null>>
];

export interface TableMeta {
  cols: Col[];
}

export default createContext<TableContextValue>([null, () => {}]);
