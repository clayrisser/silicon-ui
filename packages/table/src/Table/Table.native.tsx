import React, { FC } from 'react';
import { styled } from 'native-theme-ui';
import {
  Table as NativeTable,
  TableProps as NativeTableProps
} from 'react-native-table-component';
import {
  background,
  border,
  color,
  compose,
  layout,
  position,
  shadow,
  space,
  typography
} from 'styled-system';
import {
  StyledTableProps,
  TableProps,
  antiForwardTablePropsKeys,
  splitProps
} from './tableProps';

const StyledTable = styled<StyledTableProps, NativeTableProps>(NativeTable, {
  forwardPropsBlacklist: antiForwardTablePropsKeys
})(
  compose(
    background,
    border,
    color,
    layout,
    position,
    shadow,
    space,
    typography
  )
);

const Table: FC<TableProps> = (props: TableProps) => {
  const { customTableProps, nativeTableProps, styledTableProps } = splitProps(
    props
  );
  return (
    <StyledTable
      {...styledTableProps}
      {...nativeTableProps}
      {...customTableProps}
    />
  );
};

Table.defaultProps = {};

export default Table;
