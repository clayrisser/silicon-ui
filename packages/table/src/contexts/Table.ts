import { createContext, SetStateAction, Dispatch } from 'react';
import { Col } from '../types';

export type TableContextValue = [
  TableMeta | null,
  Dispatch<SetStateAction<TableMeta | null>>
];

export interface TableMeta {
  cols: Col[];
  resizable?: boolean;
}

export default createContext<TableContextValue>([null, () => {}]);
