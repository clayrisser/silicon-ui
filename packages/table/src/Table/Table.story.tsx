import React, { useRef } from 'react';
import { withKnobs } from '@storybook/addon-knobs';
import { Box } from '@silicon-ui/atoms/lib';
import Cell from '../Cell';
import Row from '../Row';
import Table from './Table';
import Wrapper from '../../storybook/Wrapper';
import storiesOf from '../../storybook/storiesOf';
import withThemeProvider from '../../storybook/withThemeProvider';

storiesOf('Table', module)
  .addDecorator(withKnobs)
  .addDecorator(withThemeProvider)
  .addParameters({
    // docs: { page: docs },
    // jest: ['TableHead.spec.tsx']
  })
  .add('chrome debugger', () => {
    const tableRef = useRef<any>();

    return (
      <Wrapper>
        <Box>
          <Table width="100%" resizable ref={tableRef}>
            <Row width="100%">
              <Cell
                backgroundColor="white"
                borderLeftWidth={1}
                borderTopWidth={1}
                height={200}
                width="20%"
              >
                One
              </Cell>
              <Cell
                backgroundColor="white"
                borderTopWidth={1}
                borderLeftWidth={1}
                width="20%"
                height={200}
                style={{ whiteSpace: 'normal' }}
              >
                Two three four five
              </Cell>
              <Cell
                borderTopWidth={1}
                borderLeftWidth={1}
                backgroundColor="white"
                height={200}
              >
                Three
              </Cell>
            </Row>
            <Row width="100%">
              <Cell
                backgroundColor="white"
                borderTopWidth={1}
                borderLeftWidth={1}
                borderBottomWidth={1}
                width="20%"
                height={200}
              >
                Four
              </Cell>
              <Cell
                backgroundColor="white"
                borderTopWidth={1}
                borderLeftWidth={1}
                borderBottomWidth={1}
                width="20%"
                height={200}
              >
                Five
              </Cell>
              <Cell
                backgroundColor="white"
                borderTopWidth={1}
                borderLeftWidth={1}
                borderBottomWidth={1}
                height={200}
              >
                Six
              </Cell>
            </Row>
          </Table>
        </Box>
      </Wrapper>
    );
  });
