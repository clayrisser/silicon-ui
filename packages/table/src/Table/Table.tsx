import React, { ReactNode, useState, forwardRef, LegacyRef } from 'react';
import styled, { StyledComponent } from '@emotion/styled';
import {
  background,
  border,
  color,
  compose,
  layout,
  position,
  shadow,
  space,
  typography
} from 'styled-system';
import TableContext, { TableMeta } from '../contexts/Table';
import { TableProps, DetailedHTMLTableProps, splitProps } from './tableProps';

const HTMLTable: StyledComponent<
  DetailedHTMLTableProps,
  TableProps,
  object
> = styled.table(
  compose(
    background,
    border,
    color,
    layout,
    position,
    shadow,
    space,
    typography
  )
);

const Table = forwardRef((props: TableProps, tableRef: LegacyRef<any>) => {
  const [table, setTable] = useState<TableMeta | null>({
    resizable: props.resizable,
    cols: []
  });
  const { customTableProps, styledTableProps, nativeItemProps } = splitProps(
    props
  );

  function renderRows() {
    let { children } = customTableProps;
    if (!Array.isArray(children)) children = [children];
    return ((children as unknown) as ReactNode[]).map(
      (tableRow: ReactNode) => tableRow
    );
  }

  return (
    <HTMLTable
      {...styledTableProps}
      {...nativeItemProps}
      {...(customTableProps as any)}
      style={{
        borderCollapse: 'collapse',
        tableLayout: 'fixed'
      }}
      ref={tableRef}
    >
      <TableContext.Provider value={[table, setTable]}>
        {renderRows()}
      </TableContext.Provider>
    </HTMLTable>
  );
});

Table.defaultProps = {
  backgroundColor: 'transparent',
  fontSize: 2,
  fontWeight: 'body',
  lineHeight: 'body'
};

export default Table;
