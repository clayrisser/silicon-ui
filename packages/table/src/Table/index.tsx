import Table from './Table';
import { TableProps } from './tableProps';

export { TableProps };
export default Table;
