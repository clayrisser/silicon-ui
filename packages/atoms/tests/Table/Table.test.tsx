import React from 'react';
// import { render, fireEvent } from '@testing-library/react';
import renderer from 'react-test-renderer';
import Table from '../../src/Table';

describe('<Table />', () => {
  it('renders correctly', () => {
    const tree = renderer.create(<Table />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
