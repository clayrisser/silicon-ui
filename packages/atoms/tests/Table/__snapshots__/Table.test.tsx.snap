// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Table /> renders correctly 1`] = `
<table
  className="css-1ki3w0z"
  fontSize={2}
  fontWeight="body"
  style={
    Object {
      "borderCollapse": "collapse",
    }
  }
/>
`;
